From: Conke Hu <conke.hu@gmail.com>
Date: Tue, 19 Dec 2006 21:11:37 +0000 (-0800)
Subject: PCI: ATI sb600 sata quirk
X-Git-Tag: v2.6.20~490^2~5
X-Git-Url: http://git.kernel.org/?p=linux%2Fkernel%2Fgit%2Ftorvalds%2Flinux-2.6.git;a=commitdiff_plain;h=ab17443a3df35abe4b7529e83511a591aa7384f3

PCI: ATI sb600 sata quirk

Acked-by: Jeff Garzik <jeff@garzik.org>
Cc: Alan Cox <alan@lxorguk.ukuu.org.uk>
Signed-off-by: Andrew Morton <akpm@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---

Backported to Debian's 2.6.18 by dann frazier <dannf@hp.com>

diff -urpN linux-source-2.6.18.orig/drivers/pci/quirks.c linux-source-2.6.18/drivers/pci/quirks.c
--- linux-source-2.6.18.orig/drivers/pci/quirks.c	2007-05-23 08:49:58.000000000 -0600
+++ linux-source-2.6.18/drivers/pci/quirks.c	2007-08-03 01:27:43.634528290 -0600
@@ -840,6 +840,23 @@ static void __init quirk_disable_pxb(str
 DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_INTEL,	PCI_DEVICE_ID_INTEL_82454NX,	quirk_disable_pxb );
 
 
+static void __devinit quirk_sb600_sata(struct pci_dev *pdev)
+{
+	/* set sb600 sata to ahci mode */
+	if ((pdev->class >> 8) == PCI_CLASS_STORAGE_IDE) {
+		u8 tmp;
+
+		pci_read_config_byte(pdev, 0x40, &tmp);
+		pci_write_config_byte(pdev, 0x40, tmp|1);
+		pci_write_config_byte(pdev, 0x9, 1);
+		pci_write_config_byte(pdev, 0xa, 6);
+		pci_write_config_byte(pdev, 0x40, tmp);
+
+		pdev->class = 0x010601;
+	}
+}
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_ATI, PCI_DEVICE_ID_ATI_IXP600_SATA, quirk_sb600_sata);
+
 /*
  *	Serverworks CSB5 IDE does not fully support native mode
  */
